- Description: dasd: open_count problems when using dasdfmt.
  Symptom:     dasdfmt reports 'Disk in use' when trying to format a device.
  Problem:     Usage of the open_count changed.
  Solution:    Initialize the open_count with '-1'. This means that an online
               device has an open_count of '0' and a device opened once
               (e.g. for formatting) get '1'. - as it was on kernel 2.4.
  Problem-ID:  --
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/block/dasd.c	20 Apr 2004 13:28:03 -0000	1.137.2.2
+++ linux-2.5/drivers/s390/block/dasd.c	20 Apr 2004 13:49:52 -0000	1.137.2.3
@@ -7,7 +7,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999-2001
  *
- * $Revision: 1.137.2.2 $
+ * $Revision: 1.137.2.3 $
  */
 
 #include <linux/config.h>
@@ -74,6 +74,8 @@
 	if (device == NULL)
 		return ERR_PTR(-ENOMEM);
 	memset(device, 0, sizeof (struct dasd_device));
+	/* open_count = 0 means device online but not in use */
+	atomic_set(&device->open_count, -1); 
 
 	/* Get two pages for normal block device operations. */
 	device->ccw_mem = (void *) __get_free_pages(GFP_ATOMIC | GFP_DMA, 1);
@@ -1880,7 +1882,7 @@
 	 * the blkdev_get in dasd_scan_partitions. We are only interested
 	 * in the other openers.
 	 */
-	max_count = device->bdev ? 1 : 0;
+	max_count = device->bdev ? 0 : -1;
 	if (atomic_read(&device->open_count) > max_count) {
 		printk (KERN_WARNING "Can't offline dasd device with open"
 			" count = %i.\n",

